### ISSUE TITLE
FromSql for &str?

### ISSUE THREAD

**Issue by vpzomtrrfrt** [rusqlite/rusqlite#828]
I see that `get_raw` returns a ValueRef, which can then be used to get a string reference, but could this be available more directly? Using the raw method is especially inconvenient inside query_map, since it returns a different error type

**Comment by thomcc** [comment]
Yeah, I've wanted to improve this for a while. When I was at moz we had to do awkward stuff like https://github.com/mozilla/application-services/blob/main/components/places/src/db/db.rs#L286-L306. Not great.

I have a few ideas of better APIs here (using `get` wouldn't work directly but there are ways forward), I'll see if I have time to next week.

**Comment by gwenn** [comment]
Which version of `rusqlite` is used ?
Current version of [get_raw](https://docs.rs/rusqlite/0.24.1/rusqlite/struct.Row.html#method.get_raw) panics (no `Result`).
And [get_raw_checked](https://docs.rs/rusqlite/0.24.1/rusqlite/struct.Row.html#method.get_raw_checked) doesn't panic but returns a `Result<ValueRef<'_>>`.

Also `query_map` cannot return a value with a limited lifetime (i.e. the lifetime of ValueRef).
So you must transform the `ValueRef` to X. And the best way to do it is by implementing `FromSql` for X.
If you do so, you don't need `get_raw_checked` anymore because you can use [get](https://docs.rs/rusqlite/0.24.1/rusqlite/struct.Row.html#method.get) directly.

**Comment by vpzomtrrfrt** [comment]
I want to process the &str in a way that returns a String, creating another type here would be awkward

**Comment by thomcc** [comment]
As I see it, there are two reasnable options.

- Add methods like `.get_str`/`.get_bytes` for returning `&'a str` and `&'a [u8]`.
- Add a `FromSqlRaw<'a>` trait that is implemented for `&'a str`, &'a [u8]`... etc, and add methods supporting this.

I think the 2nd is worse (I don't know how warranted the complexity is) but is still viable.

Neither of these are ideal but they would keep existing code working and simple while allowing us to improve the ergonomics of getting a borrowed output.

**Comment by gwenn** [comment]
Maybe :
```rust
impl Row<_> {
    pub fn map<I: RowIndex, F, T: FromSql>(&self, idx: I, f: F) -> Result<T>
        where F: Fn(ValueRef<'_>) -> FromSqlResult<T>,
    {
        // ... like `get` but replace `FromSql::column_result(value)` by `f(value)`
    }
}
```
?

**Comment by thomcc** [comment]
@gwenn This sort of defeats the purpose since you have to produce an owned value — the reason for using `get_raw` is that you might want to avoid copying the `&str` to a `String`.

**Comment by gwenn** [comment]
> @gwenn the reason for using `get_raw` is that you might want to avoid copying the `&str` to a `String`.

Ok, but you cannot avoid returning an owned (sub)String from `query_map`:
> Using the raw method is especially inconvenient inside query_map, since it returns a different error type

Currently, you should already be able to do something like:
```rust
let str = row.get_raw_checked(i)?.as_str()?; // auto convert FromSqlError to Error
let owned_substr = &str[1..].to_owned();
```

**Comment by thomcc** [comment]
Part of the problem is "auto convert FromSqlError to Error" isn't possible IIRC

**Comment by gwenn** [comment]
You mean https://github.com/rusqlite/rusqlite/blob/master/src/error.rs#L184-L204 does not work ?

**Comment by thomcc** [comment]
Huh, nevermind. I'll take another look later but it's possible I'm mistaken. It also might have only been an issue on Context::get_raw.

**Comment by gwenn** [comment]
https://github.com/rusqlite/rusqlite/pull/838/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R1626

So `get_raw` is not so painful to used...

**Comment by evbo** [comment]
@gwenn thank you for sharing. I have `rust-postgres` code that performs `FromSql` for `&str` and `Option<&str>`, so I did not know `get_ref()` existed. This is a total game changer, thank you for the awesome library!
